File: sdks/node/src/subject-protocol.js
import { createHash } from "node:crypto";

import { canonicalBytes } from "./encoding.js";

export class ManagedError extends Error {
  constructor(code, message, retryable = false) {
    super(message);
    this.name = "ManagedError";
    this.code = code;
    this.retryable = retryable;
  }
}

export function schemaInvalid(
  message = "The value does not satisfy the schema.",
) {
  return new ManagedError("SCHEMA_INVALID", message);
}

export function sameKeys(value, keys) {
  return (
    value !== null &&
    typeof value === "object" &&
    !Array.isArray(value) &&
    Object.keys(value).sort().join("\u0000") ===
      [...keys].sort().join("\u0000")
  );
}

export function digestBytes(value) {
  return `sha256:${createHash("sha256").update(value).digest("hex")}`;
}

export function canonicalDigest(value) {
  return digestBytes(canonicalBytes(value));
}

export function validCapability(value) {
  return (
    typeof value === "string" &&
    value.length >= 1 &&
    value.length <= 128 &&
    /^[a-z][a-z0-9.-]*$/u.test(value)
  );
}

File: sdks/python/reproit_sdk/subject_protocol.py
"""Thin validation and digest rules for subject discovery."""

from __future__ import annotations

import hashlib

from .encoding import canonical_bytes


class ManagedError(Exception):
    """Subject discovery failed with one stable local error code."""

    def __init__(self, code: str, message: str, retryable: bool = False):
        super().__init__(message)
        self.code = code
        self.message = message
        self.retryable = retryable


def schema_invalid(
    message: str = "The value does not satisfy the schema.",
) -> ManagedError:
    return ManagedError("SCHEMA_INVALID", message)


def digest_bytes(value: bytes) -> str:
    return f"sha256:{hashlib.sha256(value).hexdigest()}"


def canonical_digest(value: object) -> str:
    return digest_bytes(canonical_bytes(value))


def valid_capability(value: object) -> bool:
    if not isinstance(value, str) or not value or len(value) > 128:
        return False
    if not "a" <= value[0] <= "z":
        return False
    return all(
        "a" <= character <= "z"
        or "0" <= character <= "9"
        or character in ".-"
        for character in value[1:]
    )
